Validate annotations that refer to binaries

c3d opened the pull request:
This fixes #3004 and https://bugs.launchpad.net/katacontainers.io/+bug/1878234.

Need to add a link to code reviews once this is merged.

Forward port is https://github.com/kata-containers/kata-containers/pull/902

c3d commented:
/test

fidencio commented:
/test

c3d commented:
/test

jodh-intel reviewed (approved):
Thanks @c3d !

c3d commented:
Rebased on recent master to fix conflict

c3d commented:
/retest

c3d commented:
One more iteration required to add some variables that were missed in _makefile: Add missing generated vars to `USER_VARS`_

c3d commented:
Found a typo in `PROXYPATH` variable, was spelled `PROXYPAT`. Fixed.

c3d commented:
/retest

jodh-intel commented:
@c3d - `make --warn-undefined-variables` (which actually shows a few others in the current `master Makefile`!). Also, I found https://github.com/mrtazz/checkmake which looks interesting.

c3d commented:
Last failures apparently need rebasing on master. Pushed that.

c3d commented:
Grumble, one test had been copy-pasted, and so the removal of `vc.MinHypervisorMemory` was not acted in the added code. Retrying.

c3d commented:
/retest

c3d commented:
/retest

fidencio commented:
@c3d, I'm seeing this error accross all the CIs:
```
Drop caches
/tmp/jenkins/workspace/kata-containers-runtime-opensuse-15-PR/go/src/github.com/kata-containers/runtime
Setup virtcontainers environment
Install virtcontainers
make: Entering directory '/tmp/jenkins/workspace/kata-containers-runtime-opensuse-15-PR/go/src/github.com/kata-containers/runtime/virtcontainers'
     GOBUILD  build
     GOBUILD  hook
     GOBUILD  kata-shim
make: Leaving directory '/tmp/jenkins/workspace/kata-containers-runtime-opensuse-15-PR/go/src/github.com/kata-containers/runtime/virtcontainers'
/usr/local/bin/kata-runtime
Logging kata-env information:
no service name provided
Build step 'Execute shell' marked build as failure
Performing Post build task...
Match found for :.* : True
Logical operation result is TRUE
Running script  : #!/bin/bash

export GOPATH=$WORKSPACE/go
export GOROOT="/usr/local/go"
export PATH=${GOPATH}/bin:/usr/local/go/bin:/usr/sbin:/usr/local/bin:${PATH}

cd $GOPATH/src/github.com/kata-containers/tests
.ci/teardown.sh "$WORKSPACE/artifacts"
[kata-containers-runtime-opensuse-15-PR] $ /bin/bash /tmp/jenkins546820317887890670.sh
ERROR: kata-collect-data.sh: cannot find runtime ''
/tmp/jenkins/workspace/kata-containers-runtime-opensuse-15-PR/artifacts /tmp/jenkins/workspace/kata-containers-runtime-opensuse-15-PR/go/src/github.com/kata-containers/tests
gzip: kata-collect-data_*: No such file or directory
```

Maybe it's worth trying to reproduce the issue locally, **on a VM**. If you fancy the approach, you can simply spawn a, for instance, fedora32 VM, get the tests repo, and run a `./.ci/setup.sh` from the tests top directory. It'll setup the VM for you where you have a similar environment to try to reproduce the issue.

Let us know if some help is needed.

c3d commented:
> @c3d, I'm seeing this error accross all the CIs:
> 
> ```
> Drop caches
> /tmp/jenkins/workspace/kata-containers-runtime-opensuse-15-PR/go/src/github.com/kata-containers/runtime
> Setup virtcontainers environment
> Install virtcontainers
> make: Entering directory '/tmp/jenkins/workspace/kata-containers-runtime-opensuse-15-PR/go/src/github.com/kata-containers/runtime/virtcontainers'
>      GOBUILD  build
>      GOBUILD  hook
>      GOBUILD  kata-shim
> make: Leaving directory '/tmp/jenkins/workspace/kata-containers-runtime-opensuse-15-PR/go/src/github.com/kata-containers/runtime/virtcontainers'
> /usr/local/bin/kata-runtime
> Logging kata-env information:
> no service name provided
> Build step 'Execute shell' marked build as failure
> Performing Post build task...
> Match found for :.* : True
> Logical operation result is TRUE
> Running script  : #!/bin/bash
> 
> export GOPATH=$WORKSPACE/go
> export GOROOT="/usr/local/go"
> export PATH=${GOPATH}/bin:/usr/local/go/bin:/usr/sbin:/usr/local/bin:${PATH}
> 
> cd $GOPATH/src/github.com/kata-containers/tests
> .ci/teardown.sh "$WORKSPACE/artifacts"
> [kata-containers-runtime-opensuse-15-PR] $ /bin/bash /tmp/jenkins546820317887890670.sh
> ERROR: kata-collect-data.sh: cannot find runtime ''
> /tmp/jenkins/workspace/kata-containers-runtime-opensuse-15-PR/artifacts /tmp/jenkins/workspace/kata-containers-runtime-opensuse-15-PR/go/src/github.com/kata-containers/tests
> gzip: kata-collect-data_*: No such file or directory
> ```
> 
> Maybe it's worth trying to reproduce the issue locally, **on a VM**. If you fancy the approach, you can simply spawn a, for instance, fedora32 VM, get the tests repo, and run a `./.ci/setup.sh` from the tests top directory. It'll setup the VM for you where you have a similar environment to try to reproduce the issue.
> 
> Let us know if some help is needed.

Thanks. The problem is yet another `@XYZ@` variable not being replaced. Here, it's `@RUNTIME_NAME@` in `kata-collect-data.sh`. However, this time it's not because the variable is not there, nor because the file is not in `GENERATED_VARS`. So some new failure mode. Apparently, it is being replaced, but the variable is not set while building. So we end up creating a `kata-collect.sh` that has the runtime set to `""`.

I'm currently investigating. I can reproduce in my local build, so no need for a VM at the moment, although I am likely to try and run the CI locally after that change.

c3d commented:
> I'm currently investigating. I can reproduce in my local build, so no need for a VM at the moment, although I am likely to try and run the CI locally after that change.

OK, found the problem. For some reason, this was a special case for the replacement in the original makefile, which I missed. The replacement is `RUNTIME_NAME` -> `$(TARGET)`. 

A smart Emacs grep regexp search (`s|@\([A-Z0-9_]+]\)@|\$(\1)\g` if you are curious) allowed me to see that this is the only legitimate case, but that there were one typo:

```
		-e "s|@DEFENABLEMSWAP@|$(DEFENABLESWAP)|g" \
```

So this should be the last one, hopefully, Will check if the problem exists on the 2.0 dev branch.

c3d commented:
/retest

c3d commented:
@fidencio It now looks good. @devimc, @bergwolf, @gabyct, one last quick look?

devimc commented on cli/config/configuration-qemu-virtiofs.toml.in:
```diff
@@ -16,6 +16,19 @@ kernel = "@KERNELVIRTIOFSPATH@"
+# Each member of the list is a regular expression, which is the base name
+# of the annotation, e.g. "path" for io.katacontainers.config.hypervisor.path"
+# The default if not set is empty (all annotations rejected.)
+# Your distribution recommends: @DEFENABLEANNOTATIONS@
```
q. why the comment here is different to the rest of `.toml.in` files?

```toml
# List of valid annotation names for the hypervisor
# Each member of the list is a regular expression, which is the base name
# of the annotation, e.g. "path" for io.katacontainers.config.hypervisor.path"
````

c3d replied:
Interesting. I thought I had unified all the docs. Good catch, I will align all the others to the on for virtiofs (which is more verbose, preferred).

Good point about the docs as well. I will submit a different PR for that.

c3d replied:
Ah, I take that back. There are two distinct comments, and they are consistent across files.

The `enable_annotations` configuration checks the _names_ of the annotations you will accept. For example, you may want to accept only `kernel_params` annotations, but reject an annotation that modifies the hypervisor.

A second level of check is that if you accept the hypervisor annotation, you can still restrict which values are accepted.

c3d replied:
Any suggestion for a clearer wording is welcome, BTW.

devimc replied:
I like this comment, just make it consistent across `toml.in` files (all file should have the same comments)

c3d replied:
> I like this comment, just make it consistent across `toml.in` files (all file should have the same comments)

Sorry, I somehow had missed the inconsistency, despite checking. I saw it now.

c3d replied:
I will make it a separate commit so that I can apply on 2.0 where the changes were already merged.

devimc commented on cli/config/configuration-qemu-virtiofs.toml.in:
```diff
@@ -16,6 +16,19 @@ kernel = "@KERNELVIRTIOFSPATH@"
+# Your distribution recommends: @DEFENABLEANNOTATIONS@
+enable_annotations = @DEFENABLEANNOTATIONS@
+
+# List of valid annotations values for the hypervisor
```
sorry this is confusing me:  _List of valid **annotations** values for the hypervisor_  - or -   _List of **paths** for the hypervisor_ ? :thinking: 
an annotation look like this: "io.kata.feature", a path looks like this: "/usr/bin/qemu"

c3d replied:
An annotation looks like `io.kata.feature-name: feature-value`. These are called "name-value" pairs, so I refer to the value for the annotation here.

Does that make sense?

(The fact that the value is a path here is relevant, so I may change the wording to reflect that).

c3d replied:
> (The fact that the value is a path here is relevant, so I may change the wording to reflect that).

However, re-reading, I'm not sure how to improve the wording. Any suggestion?

devimc replied:
might be  _List of paths for the hypervisor_   ..?

c3d replied:
Updated the comments in a separate commit in order to be able to forward-port to 2.0 where this is already merged.

c3d replied:
Done

devimc reviewed (commented):
thanks @c3d - I left some questions and comments

btw `(all annotations rejected.)` , that's a big change,  would you mind updating the documentations as well?

fidencio commented on Makefile:
```diff
@@ -47,6 +47,9 @@ PROJECT_ORG = $(PROJECT_TAG)
 PROJECT_BUG_URL = $(PROJECT_URL)/kata-containers/issues/new
 
+# @RUNTIME_NAME@ should be replaced with the target in generated files
+RUNTIME_NAME = $(TARGET)
```
@c3d, shouldn't `RUNTIME_NAME` be set only after `TARGET` is set?
TARGET is set a few lines below.

c3d replied:
Does not really matter if you assign it with `=`. In that case, it's re-evaluated when you use it.

fidencio replied:
As you pointed out in the PR for 2.0.0, at the first look it's confusing, indeed.
There's no need to change it, unless you want to / will already rework the series for some other reason.

Thanks for the explanation!

fidencio reviewed (commented):
@c3d, I've left a question inline.

fidencio reviewed (approved):
@c3d, lgtm, thanks!

Please, let's also wait for devimc's approval before pushing it.

jodh-intel commented:
@c3d - please could you include the commits I created on https://github.com/kata-containers/kata-containers/pull/1002/ before this lands? I've added the dnm label for now.

c3d commented:
> @c3d - please could you include the commits I created on [kata-containers/kata-containers#1002](https://github.com/kata-containers/kata-containers/pull/1002) before this lands? I've added the dnm label for now.

Looking at it.

c3d commented:
@jodh-intel  Pushed an updated version. I moved your fix up in the history, next to the breakage, to minimize risk of bisect issue. Also, I replaced `:=` with `=` to match the neighbouring assignments (see my comment on 1002)

jodh-intel commented:
Thanks @c3d.

lgtm

/test

c3d commented:
Hey @GabyCT could you take a look at the Jenkins failures here and tell me if anything looks out of the ordinary?

(Context: there is a makefile change that may have impacted replacements of `@FOO@` in the `.in` files)

GabyCT commented:
@c3d I see that there is a failure in metrics job that is consistent with this PR
```
05:49:19 opt/kata/bin/qemu-virtiofs-system-x86_64
05:49:19 time="2020-10-20T10:49:19Z" level=error msg="/usr/share/defaults/kata-containers/configuration.toml: file /usr/bin/qemu-lite-system-x86_64 does not exist" arch=amd64 name=kata-runtime pid=12891 source=runtime
05:49:19 /usr/share/defaults/kata-containers/configuration.toml: file /usr/bin/qemu-lite-system-x86_64 does not exist
```

GabyCT commented:
@c3d the same failure is in virtiofs CI, this is again related with this PR
```
05:20:01 opt/kata/bin/qemu-virtiofs-system-x86_64
05:20:01 time="2020-10-20T10:20:01Z" level=error msg="/usr/share/defaults/kata-containers/configuration.toml: file /usr/bin/qemu-lite-system-x86_64 does not exist" arch=amd64 name=kata-runtime pid=15891 source=runtime
05:20:01 /usr/share/defaults/kata-containers/configuration.toml: file /usr/bin/qemu-lite-system-x86_64 does not exist
05:20:01 Failed at 78: sudo -E PATH=$PATH "$kata_runtime_path" kata-check
```

GabyCT commented:
@c3d the same failure is in virtiofs CI, this is again related with this PR
```
05:20:01 opt/kata/bin/qemu-virtiofs-system-x86_64
05:20:01 time="2020-10-20T10:20:01Z" level=error msg="/usr/share/defaults/kata-containers/configuration.toml: file /usr/bin/qemu-lite-system-x86_64 does not exist" arch=amd64 name=kata-runtime pid=15891 source=runtime
05:20:01 /usr/share/defaults/kata-containers/configuration.toml: file /usr/bin/qemu-lite-system-x86_64 does not exist
05:20:01 Failed at 78: sudo -E PATH=$PATH "$kata_runtime_path" kata-check
```

GabyCT commented:
for the ARM CI failures, I am not familiar but @jongwu any idea of this?

jongwu commented:
@GabyCT -， arm ci fail for cpu hotplug is not supported on arm64, I have sent [PR](https://github.com/kata-containers/tests/pull/2978) to skip it.

c3d commented:
@GabyCT @jongwu Thanks a lot. I'll look into the failures that @GabyCT pointed to.

amshinde commented on virtcontainers/persist.go:
```diff
@@ -224,6 +224,7 @@ func (s *Sandbox) dumpConfig(ss *persistapi.SandboxState) {
 		MachineAccelerators:     sconfig.HypervisorConfig.MachineAccelerators,
 		CPUFeatures:             sconfig.HypervisorConfig.CPUFeatures,
 		HypervisorPath:          sconfig.HypervisorConfig.HypervisorPath,
+		HypervisorPathList:      sconfig.HypervisorConfig.HypervisorPathList,
```
I am yet to review the rest of the PR, so not sure how this is being used later.  But I do not understand why these variables are getting persisted at all. I would assume that these would just be needed at startup to validate the binary paths. And its not like they are reconfigured or changed at runtime.

c3d replied:
My experience was that I needed that for the mechanism to work in practice.

From reading the code, it looks like even container startup goes through `fetchSandbox`. I may be misunderstanding how this is supposed to work. Do you see any drawback to having the correct configuration for all other operations (besides a bit of extra memory)?

c3d replied:
Thank you for the explanation.

c3d commented:
> @c3d I see that there is a failure in metrics job that is consistent with this PR
> 
> ```
> 05:49:19 opt/kata/bin/qemu-virtiofs-system-x86_64
> 05:49:19 time="2020-10-20T10:49:19Z" level=error msg="/usr/share/defaults/kata-containers/configuration.toml: file /usr/bin/qemu-lite-system-x86_64 does not exist" arch=amd64 name=kata-runtime pid=12891 source=runtime
> 05:49:19 /usr/share/defaults/kata-containers/configuration.toml: file /usr/bin/qemu-lite-system-x86_64 does not exist
> ```

Apparently a fixup error that caused variable `QEMUVIRTIOFSPATH` to be removed mistakenly from makefile. Fixed.

c3d commented:
Not sure why I don't see the push I just made here.

/test

c3d commented:
@GabyCT It looks like http://jenkins.katacontainers.io/job/kata-containers-runtime-ubuntu-PR-cloud-hypeprvisor-minimal/ fails with 

```
14:37:22 Build timed out (after 5 minutes). Marking the build as aborted.
```

I don't think this is related to my changes. Could you try increasing the build time for that job?

c3d commented:
@GabyCT I'm sorry to bug you again, but this time the problems I see are docker failures and complaints that there are no logs. I would be surprised if this is related to the changes here. Maybe I'm not looking at the right spot?

```
14:39:35 Job for docker.service failed because the control process exited with error code.
14:39:35 See "systemctl status docker.service" and "journalctl -xe" for details.
```

Followed by

```
14:39:10 Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
14:39:10 Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
14:39:10 INFO: Runtime environment not found - setting defaults
14:39:10 INFO: delete stale docker resource under /var/lib/docker
14:39:10 INFO: delete stale kata resource under /var/lib/vc /run/vc /usr/share/kata-containers /usr/share/defaults/kata-containers
14:39:10 INFO: Remove installed kata packages
```

At the end:

```
14:39:38 ERROR: kata-log-parser: no log records to process
```

But that seems to apply to components I did not touch at all, e.g  proxy or shim.

c3d commented:
@GabyCT When testing in a VM locally, I have the same failure. `journalctl -xe` tells me:

```
Oct 21 17:18:28 f32-turbo dockerd[20927]: time="2020-10-21T17:18:28.290087228+02:00" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc42044b940, CONNECTING" module=grpc
Oct 21 17:18:28 f32-turbo dockerd[20927]: time="2020-10-21T17:18:28.290300094+02:00" level=info msg="pickfirstBalancer: HandleSubConnStateChange: 0xc42044b940, READY" module=grpc
Oct 21 17:18:28 f32-turbo dockerd[20927]: time="2020-10-21T17:18:28.292459046+02:00" level=info msg="[graphdriver] using prior storage driver: overlay2"
Oct 21 17:18:28 f32-turbo dockerd[20927]: time="2020-10-21T17:18:28.549707306+02:00" level=info msg="Graph migration to content-addressability took 0.00 seconds"
Oct 21 17:18:28 f32-turbo dockerd[20927]: time="2020-10-21T17:18:28.550244601+02:00" level=warning msg="Your kernel does not support cgroup memory limit"
Oct 21 17:18:28 f32-turbo dockerd[20927]: time="2020-10-21T17:18:28.550324811+02:00" level=warning msg="Unable to find cpu cgroup in mounts"
Oct 21 17:18:28 f32-turbo dockerd[20927]: time="2020-10-21T17:18:28.550396035+02:00" level=warning msg="Unable to find blkio cgroup in mounts"
Oct 21 17:18:28 f32-turbo dockerd[20927]: time="2020-10-21T17:18:28.550463190+02:00" level=warning msg="Unable to find cpuset cgroup in mounts"
Oct 21 17:18:28 f32-turbo dockerd[20927]: time="2020-10-21T17:18:28.550591721+02:00" level=warning msg="mountpoint for pids not found"
Oct 21 17:18:28 f32-turbo dockerd[20927]: Error starting daemon: Devices cgroup isn't mounted
Oct 21 17:18:28 f32-turbo systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
-- Subject: Unit process exited
```

This is with a clean Fedora32 VM. Is there some missing `cgroup` setup in the CI setup script?

GabyCT commented:
@c3d the Fedora 32 CI has `docker` and it was a custom image where we enable the cgroups to support `docker` as by defaul it supports `podman`

GabyCT commented:
@c3d looking the logs of the failures, I see the following
```

Output of "`/usr/bin/kata-runtime kata-env`":
```toml
Cannot find usable config file (config file "/etc/kata-containers/configuration.toml" unresolvable: file /etc/kata-containers/configuration.toml does not exist, config file "/usr/share/defaults/kata-containers/configuration.toml" unresolvable: file /usr/share/defaults/kata-containers/configuration.toml does not exist)
```

c3d commented:
@GabyCT I'm a bit of a newbie with respect to the CI. I've looked at 3 Jenkins jobs already, did not see that message. Could you give me a link to the log you are looking at?

GabyCT commented:
@c3d no problem, to see the logs what I did was the following
1. go to the failure job http://jenkins.katacontainers.io/job/kata-metrics-runtime-ubuntu-virtiofs-18-04-PR/131/
2. go to `build artifacts` you will find the `kata-collect` logs and there is where I found the failure

c3d commented:
> @c3d looking the logs of the failures, I see the following
> 
> ```
> 
> Output of "`/usr/bin/kata-runtime kata-env`":
> ```toml
> Cannot find usable config file (config file "/etc/kata-containers/configuration.toml" unresolvable: file /etc/kata-containers/configuration.toml does not exist, config file "/usr/share/defaults/kata-containers/configuration.toml" unresolvable: file /usr/share/defaults/kata-containers/configuration.toml does not exist)
> ```

@GabyCT I tried that on `master` as well, I see no difference in behavior here. Are you sure this is the cause of the failure?

c3d commented:
@GabyCT While looking at the same logs trying to find something odd, I noticed this:

```
## version

Output of "` --version`":
```
/usr/bin/kata-collect-data.sh: line 178: --version: command not found
```

This seems to come from the `throttlers` variable not being set in https://github.com/kata-containers/runtime/blob/master/data/kata-collect-data.sh.in#L592.  Looks like a bug in the script to me. I'm trying to see if something else in this PR could have caused it indirectly.

GabyCT commented:
@c3d the metrics CI issue has been solved

c3d commented:
I just noticed that all the required tests were passing. I was confused because the Jenkins icon is red :-( No wonder I had trouble finding new failures…

fidencio commented:
/test

amshinde commented:
I'd like to take a look at this before it is merged.

fidencio commented:
@amshinde, I've added the do-not-merge label to avoid getting it merged by accident. Please, feel free to remove it once you're done with the reviews.

Also, thanks for taking a look at the series!

amshinde commented on virtcontainers/pkg/annotations/annotations.go:
```diff
@@ -44,6 +44,9 @@ const (
 	JailerPath = kataAnnotHypervisorPrefix + "jailer_path"
 
+	// CtlPath is a sandbox annotation for passing a per container path pointing at the acrn ctl binary
+	CtlPath = kataAnnotHypervisorPrefix + "ctlpath"
```
Ah, see that you added this as an annotation. I would rather go in the other direction of trimming down our annotations in the future and only support those that are truly required, rather than support all configs as annotations from usability and maintenance point of view . Would like to see what others think about this.
Also, since this is a very `ACRN` specific annotation, I am not in favor of adding this unless someone explicitly requests for it to be added.

c3d replied:
This is a security risk. It is important to ensure that if the annotation is ever added, it follows the correct pattern. Otherwise, we reopen the big hole that just series closes.

fidencio replied:
@c3d, we're on the same page on how important it's to ensure that if an annotation is ever added, it must follow the correct pattern. However, I also agree with @amshinde that we should not add any new annotation unless it's stirctly needed. With this in mind, I'd like to understand how CtlPath would take advantage of an annotation.

@kata-containers/runtime, any opinion here?

amshinde commented on cli/config/configuration-acrn.toml.in:
```diff
@@ -16,6 +16,11 @@ ctlpath = "@ACRNCTLPATH@"
+# List of valid annotation names for the hypervisor
+# Each member of the list is a regular expression, which is the base name
+# of the annotation, e.g. "path" for io.katacontainers.config.hypervisor.path"
+enable_annotations = @DEFENABLEANNOTATIONS@
```
This is quite a big change to disable annotations altogether. We should discuss this in the arch call and document this properly after agreement. Also annotations will no longer work for any users making use of this feature, so I think we need to highlight this once this goes in a release.

c3d replied:
> This is quite a big change to disable annotations altogether. We should discuss this in the arch call and document this properly after agreement. Also annotations will no longer work for any users making use of this feature, so I think we need to highlight this once this goes in a release.

@amshinde This change has been discussed in the arch meeting several times already. It was discussed since May. See https://bugs.launchpad.net/katacontainers.io/+bug/1878234 and https://github.com/kata-containers/runtime-private/pull/1 to see the extent of the discussion, to which you took place. I only moved here once it was approved. However, now that the cat is out of the bag, I would like to move ahead with what was approved, since this is a security risk and a CVE, and not restart here a discussion that was help under wraps for good reasons over the past 6 months.

c3d replied:
> Another question I have is I only see this for hypervisor specific section here, so are annotations related to the runtime/agent affected by this change?

No. Annotations in the agent cannot cause the kind of damage the runtime can. See the discussion on Launchpad.

amshinde replied:
I just found it a bit inconsistent that the hypervisor related annotations are disabled while those related to runtime/agent  are enabled. Its worth making this clear in the commit message and in the PR itself.

c3d replied:
The plan is definitely to add similar control for other components. if only for consistency. However, since this does not represent a security risk, it can be deferred. We discussed that earlier.

fidencio replied:
@amshinde, would you be okay with this change is as long as a patch updating our documentation comes together with this series?

amshinde replied:
Yes, I am fine with it as long as we document it.

c3d replied:
@amshinde Could you clarify, when you talk about defaults, if you are talking about

1. The defaults from the binary if there is nothing in the configuration file  (currently empty list for everything)
2. The defaults in the `config*.toml.in` files, which currently are always some `@DEFBLAH`
3. The defaults in the makefile.

Right now, for example, the value of `DEFENABLEANNOTATIONS` in the makefile is `[]`. I could very easily be convinced to replace it with `[".*"]`, i.e. enable hypervisor annotations in our builds. Is that what you were talking about?

I'll push a commit with that single change in order to be able to forward-port it to 2.0. @kata-containers/architecture-committee , do you agree that enabling hypervisor annotations in the default configuration files is the right thing to do?

c3d replied:
@amshinde Assuming I (finally) correctly understood what you suggested, the change is in f26b3f9bf78fa12a448b3f98b5578748c2164a9d. I marked it as "suggested-by" you. If I misunderstood, let me know (and reopen this thread).

amshinde commented on cli/config/configuration-acrn.toml.in:
```diff
@@ -16,6 +16,11 @@ ctlpath = "@ACRNCTLPATH@"
+# List of valid annotation names for the hypervisor
+# Each member of the list is a regular expression, which is the base name
+# of the annotation, e.g. "path" for io.katacontainers.config.hypervisor.path"
+enable_annotations = @DEFENABLEANNOTATIONS@
```
Another question I have is I only see this for hypervisor specific section here, so are annotations related to the runtime/agent affected by this change?

c3d replied:
(Reposting comment, which for some reason I attached to your previous comment)

> Another question I have is I only see this for hypervisor specific section here, so are annotations related to the runtime/agent affected by this change?

No. Annotations in the agent cannot cause the kind of damage the runtime can. See the discussion on Launchpad.

amshinde replied:
Again, I asked because in the discussion I was under the impression that we talked about enabling/disabling all annotations and not hypervisor related annotations.

c3d replied:
In the long run, yes. It's less important and urgent, however.

c3d commented on pkg/katautils/config.go:
```diff
@@ -86,9 +86,12 @@ type factory struct {
 	JailerPath              string   `toml:"jailer_path"`
+	JailerPathList          []string `toml:"jailer_path_list"`
 	Kernel                  string   `toml:"kernel"`
 	CtlPath                 string   `toml:"ctlpath"`
```
> I dont think CtlPath can be configured through annotations, as far as I can tell this is a `ACRN` hypervisor specific configuration not exposed through annotations at all, just set it the configuration file.
> So I question if we need to validate this at all since, given only the root user can change the configuration.

Apparently the `CtlPath` annotation exists in the code:

```
	// CtlPath is a sandbox annotation for passing a per container path pointing at the acrn ctl binary
	CtlPath = kataAnnotHypervisorPrefix + "ctlpath"
```

I have not checked if this can be modified successfully through annotations today. In my opinion, it does not matter: if the annotation exists and can't be used, someone someday will consider it a bug worth fixing, and I'd rather have the correct template for them to follow already in place.

amshinde replied:
@c3d This annotation did not exist before, has been added with this commit in your PR:
https://github.com/kata-containers/runtime/pull/3005/commits/999a77e56cb211d6d67c1a3a58b95ad9267b7507

Prior to your change, this value could not be changed by an annotation.
My point is we do not want to support a config as an annotation unless required to avoid bloat. Any future annotations should be added if there is a need for it and reviewed to pass validation as you have introduced here for existing annotations.

c3d replied:
@amshinde Can you explain the rationale why we can use annotations for any binary except ctl? This looks to me more like an oversight than a carefully designed decision.

In any case, knowing the annotation security issue, the trade-off here is between security + consistency on one hand, and bloat on the other hand. So it's interesting to evaluate the bloat. The difference in binary size on my machine is 4256 bytes, on a binary that is 36M is size. That's less than 0.01% change. It's worth it to pre-emptively close a security gap.

c3d replied:
@amshinde Marking as resolved, to clarify the view. Feel free to reopen if you think more action is needed here.

c3d commented on virtcontainers/pkg/oci/utils.go:
```diff
@@ -368,7 +378,7 @@ func addAssetAnnotations(ocispec specs.Spec, config *vc.SandboxConfig) {
 }
 
-func addHypervisorConfigOverrides(ocispec specs.Spec, config *vc.SandboxConfig) error {
+func addHypervisorConfigOverrides(ocispec specs.Spec, config *vc.SandboxConfig, runtime RuntimeConfig) error {
```
Actually, I had used `runtime` because that is what was used elsewhere in the code, e.g. in `SandboxConfig`.

fidencio replied:
I feel @amshinde is referring to other pieces of the code where we deal with the runtime config. A grep shows me those are indeed treated as "runtimeConfig".

@amshinde pointed is as a nit, and although I have a slightly preference for using the suggested `runtimeConfig` I don't a strong opinion on this.

The only thing I may ask is, if not changing (which is perfectly fine), please, make this as resolved.

c3d replied:
Changed in a separate commit e61f02d583ccab14f0de393cf0e9361a656d564c, also changed the existing usage in `SandboxConfig` function to be consistent across the file.

c3d commented:
@amshinde are you still reviewing this? If you are done, could you please remove the `do-not-merge` flag?

c3d commented:
Pushed a couple of comment-only commits. Revalidating Jenkins tests.

/test

c3d commented on cli/config/configuration-acrn.toml.in:
```diff
@@ -16,6 +16,10 @@ ctlpath = "@ACRNCTLPATH@"
 kernel = "@KERNELPATH_ACRN@"
 image = "@IMAGEPATH@"
 
+# List of valid annotations values for the hypervisor (default: empty)
```
Fixed in the last iteration. @devimc had made a similar remark.

c3d commented on cli/config/configuration-acrn.toml.in:
```diff
@@ -16,6 +16,10 @@ ctlpath = "@ACRNCTLPATH@"
 kernel = "@KERNELPATH_ACRN@"
 image = "@IMAGEPATH@"
 
+# List of valid annotations values for the hypervisor (default: empty)
```
You have to distinguish
* The runtime default if given nothing in the configuration, which is empy. This is the only safe value.
* The disto default which, unless overriden by a specific distro,  is to have the path to the default location.

In other words, there is a sane default in the config file, which is not empty, and a paranoid default in the runtime, which is that if the config does not have an entry, the security issue is closed.

fidencio replied:
This is, IMHO, the same case as https://github.com/kata-containers/runtime/pull/3005#discussion_r512219321.

fidencio commented on cli/config/configuration-clh.toml.in:
```diff
@@ -62,6 +62,9 @@ default_memory = @DEFMEMSZ@
 virtio_fs_daemon = "@DEFVIRTIOFSDAEMON@"
 
+# List of valid annotations values for the virtiofs daemon (default: empty)
+# virtio_fs_daemon_list = [ "/opt/kata/bin/virtiofsd", "/usr/.*/virtiofsd" ]
```
@amshinde, I know this question may deviate a little bit from your requestm but the defaults on Kata should be the ones optmized for the kata-deploy usecase? If so, I'd say that, by default, we just allow the paths known and used by kata deploy.

Does this sound reasonable @amshinde? @c3d, what do you think?

c3d commented:
/test

devimc reviewed (approved):
lgtm - thanks @c3d

jodh-intel commented:
@c3d - This branch is now conflicted (due to #3031 landing). The fix is relatively easy - you'll need to tweak this PR in 2 places I believe:

- Update `addAnnotations()` to deal with the new return value of `addAssetAnnotations()`:

  https://github.com/kata-containers/kata-containers/pull/1086/files#diff-bf17dd8525e740a4d01fafbd5f48a42b621cf32540465de2e14c6b15ab1124deL362-R365
- Update `TestAddAssetAnnotations()` to create the `fakeAssetFile` file:

  https://github.com/kata-containers/kata-containers/pull/1086/files#diff-7abc7cc1240550373976626143dcee3341b4630324070be96a7ff13f9aaa1624R662-R693

Related: https://github.com/kata-containers/kata-containers/pull/1086 which already includes [the 2.x version of this PR](https://github.com/kata-containers/kata-containers/pull/902).

c3d commented:
> @c3d - This branch is now conflicted (due to #3031 landing). The fix is relatively easy - you'll need to tweak this PR in 2 places I believe:

Fix pushed.

fidencio commented:
@amshinde, are we fine to have this one merged? :-)

jodh-intel commented:
@c3d - thanks, but I don't think you updated `TestAddAssetAnnotations()` as mentioned above? The CI will fail without the 2 changes I mentioned I believe.

c3d commented:
> @c3d - thanks, but I don't think you updated `TestAddAssetAnnotations()` as mentioned above? The CI will fail without the 2 changes I mentioned I believe.

Ooops, I must be tired. Did a git push without committing the changes…

jodh-intel reviewed (approved):
Thanks @c3d.

jodh-intel commented:
/test

amshinde commented:
@c3d Test `TestCheckPathIsInGlobs` is failing. Can you take a look?

c3d commented:
> @c3d Test `TestCheckPathIsInGlobs` is failing. Can you take a look?

I need to think a bit about this one. Replacing actual paths with a file called `fake-binary` now fails, and that is correct, since the binary does not exist. That's one difference between `glob()` and regular expressions. I'm need to re-think the test to work in that new context.

c3d commented:
> > @c3d Test `TestCheckPathIsInGlobs` is failing. Can you take a look?
> 
> I need to think a bit about this one. Replacing actual paths with a file called `fake-binary` now fails, and that is correct, since the binary does not exist. That's one difference between `glob()` and regular expressions. I'm need to re-think the test to work in that new context.

I was wrong. That has been taken care of. It is failing for another reason, which is easy to fix.

jodh-intel commented:
@c3d - it looks like you may have just missed off the following part of my PR, specifically to set the `runtimeConfig.HypervisorConfig.*PathList` annotations?:

https://github.com/kata-containers/kata-containers/pull/1086/files#diff-7abc7cc1240550373976626143dcee3341b4630324070be96a7ff13f9aaa1624R726-R733

fidencio commented:
/test

c3d commented:
@fidencio I pushed a rebased version. Let's try to figure out the CI issues.

c3d commented:
/test

fidencio commented:
- **jenkins-ci-podman-clh-fedora-31**
This is the issue happening:
```
18:18:23 Note: switching to 'v0.11.0'.
18:18:23 
18:18:23 You are in 'detached HEAD' state. You can look around, make experimental
18:18:23 changes and commit them, and you can discard any commits you make in this
18:18:23 state without impacting any branches by switching back to a branch.
18:18:23 
18:18:23 If you want to create a new branch to retain commits you create, you may
18:18:23 do so (now or later) by using -c with the switch command. Example:
18:18:23 
18:18:23   git switch -c <new-branch-name>
18:18:23 
18:18:23 Or undo this operation with:
18:18:23 
18:18:23   git switch -
18:18:23 
18:18:23 Turn off this advice by setting config variable advice.detachedHead to false
18:18:23 
18:18:23 HEAD is now at 7dcd3aff build, release-notes.md: Document 0.11.0 release
18:18:23 ./scripts/dev_cli.sh: line 122: docker: command not found
18:18:23 [Cloud Hypervisor] Error pulling container image. Aborting.
18:18:23 Failed at 41: "${GOPATH}/src/${packaging_repo}/static-build/cloud-hypervisor/build-static-clh.sh"
18:18:23 Failed at 54: "${cidir}/install_cloud_hypervisor.sh"
18:18:23 ++ handle_error 69
18:18:23 ++ local exit_code=127
18:18:23 ++ local line_number=69
18:18:23 ++ echo 'Failed at /tmp/jenkins/workspace/kata-containers-runtime-podman-clh-fedora-31-PR/ci_entry_point.sh +69: .ci/jenkins_job_build.sh "${repo_to_test}"'
18:18:23 Failed at /tmp/jenkins/workspace/kata-containers-runtime-podman-clh-fedora-31-PR/ci_entry_point.sh +69: .ci/jenkins_job_build.sh "${repo_to_test}"
18:18:23 ++ exit 127
```
The issue happens because we don't install docker in this CI (we're testing podman there), but cloud-hypervisor explicitly requires docker: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/scripts/dev_cli.sh#L13

Surprises me if this test has ever passed.  There are two ways of solving this issue:
1. Simply do a  `ln -sf /usr/bin/podman /usr/bin/docker` on our side;
2. Teach cloud-hypervisor to deal with podman;

Both are solutions that should be applied, and I'll start with the simple hack on our side.

fidencio commented:
- **jenkins-ci-ubuntu-1804-cloud-hypervisor-k8s-e2e-containderd-minimal**:
```
18:31:19 running: sonobuoy run --wait=180 --e2e-focus="ConfigMap should be consumable from pods in volume|InitContainer \[NodeConformance\] should invoke init containers|Kubelet when scheduling a busybox command in a pod should printk|Projected secret should be consumable in multiple volumes|Secrets should be consumable via the environment|\[sig-storage\] Downward API volume should update annotations on modification \[NodeConformance\] \[Conformance\]"
18:31:19 time="2020-11-10T17:31:19Z" level=info msg="created object" name=sonobuoy namespace= resource=namespaces
18:31:19 time="2020-11-10T17:31:19Z" level=info msg="created object" name=sonobuoy-serviceaccount namespace=sonobuoy resource=serviceaccounts
18:31:19 time="2020-11-10T17:31:19Z" level=info msg="created object" name=sonobuoy-serviceaccount-sonobuoy namespace= resource=clusterrolebindings
18:31:19 time="2020-11-10T17:31:19Z" level=info msg="created object" name=sonobuoy-serviceaccount-sonobuoy namespace= resource=clusterroles
18:31:19 time="2020-11-10T17:31:19Z" level=info msg="created object" name=sonobuoy-config-cm namespace=sonobuoy resource=configmaps
18:31:19 time="2020-11-10T17:31:19Z" level=info msg="created object" name=sonobuoy-plugins-cm namespace=sonobuoy resource=configmaps
18:31:20 time="2020-11-10T17:31:20Z" level=info msg="created object" name=sonobuoy namespace=sonobuoy resource=pods
18:31:20 time="2020-11-10T17:31:20Z" level=info msg="created object" name=sonobuoy-master namespace=sonobuoy resource=services
18:36:20 Build timed out (after 5 minutes). Marking the build as aborted.
18:36:20 Build was aborted
```
@amshinde, @jcvenegas, @devimc, have you seen this one before? Any tip of what it may be related to?

fidencio commented:
- **jenkins-ci-Power8-ubuntu-18-04-initrd**:
```
Install os-tree
Job for docker.service failed because the control process exited with error code.
See "systemctl status docker.service" and "journalctl -xe" for details
```
@bpradipt, @Amulyam24 @clnperez, @jing-wang4, do you have some idea of what may be the issue here?

fidencio commented:
- **jenkins-ci-opensuse-15**
Failing while performing a `yum update` due to:
```
error: unpacking of archive failed on file /usr/bin/newgidmap;5faad0a7: cpio: cap_set_file failed - Inappropriate ioctl for device
error: shadow-utils-2:4.6-9.fc30.x86_64: install failed
```

This is our good and old issue about not having `-o xattrs` passed to virtiofsd.
My suggestion is to **always** add that to our CI, regardless of the security implications, till we have a proper fix upstream.  By doing this we'll be able to unblock CI.

fidencio commented:
- **jenkins-ubuntu-18-04-vfio**:
@devimc, I'm not capable to look at the errors and figure out whether it's expected or not. May I borrow your brain on this one? :-)

devimc commented:
vfio CI failed testing clh, restarting CI

likebreath commented:
@fidencio You are right that the `jenkins-ci-podman-clh-fedora-31` CI is an known issue (https://github.com/kata-containers/ci/issues/340).  I think we are fine to move on with this CI failure as it is unrelated to this PR. (@jcvenegas had a similar idea as your `solution 1` and was experimenting it in a debugging PR https://github.com/kata-containers/tests/pull/2922)

Also, I noticed the other two clh failing CIs were both passing yesterday: clh-k8s-containerd job [#159](http://jenkins.katacontainers.io/job/kata-containers-runtime-ubuntu-PR-cloud-hypeprvisor-k8s-containerd/159/) and clh-k8s-e2e-containerd-mini job [#164](http://jenkins.katacontainers.io/job/kata-containers-runtime-ubuntu-PR-cloud-hypeprvisor-k8s-e2e-containerd-minimal/164/). So I just restarted those two CI jobs to see whether we got sporadic failures.

amshinde commented:
for Firecracker, most of the CI has passed, but failed at the last step at :
```
END OF POST BUILD TASK : 0
Archiving artifacts
ERROR: Step ‘Archive the artifacts’ aborted due to exception: 
java.io.IOException: No space left on device
```

We can ignore this as this is an unrelated error. @chavafg Have you seen this before?

fidencio commented:
- **jenkins-ci-ubuntu-1804-cloudhypervisor-docker**:
```
19:28:55 POST BUILD TASK : FAILURE
19:28:55 END OF POST BUILD TASK : 0
19:28:55 Archiving artifacts
19:29:00 ERROR: Step ‘Archive the artifacts’ aborted due to exception: 
19:29:00 java.io.IOException: No space left on device
19:29:00 	at sun.nio.ch.FileDispatcherImpl.write0(Native Method)
19:29:00 	at sun.nio.ch.FileDispatcherImpl.write(FileDispatcherImpl.java:60)
19:29:00 	at sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:93)
19:29:00 	at sun.nio.ch.IOUtil.write(IOUtil.java:65)
19:29:00 	at sun.nio.ch.FileChannelImpl.write(FileChannelImpl.java:211)
19:29:00 	at java.nio.channels.Channels.writeFullyImpl(Channels.java:78)
19:29:00 	at java.nio.channels.Channels.writeFully(Channels.java:101)
19:29:00 	at java.nio.channels.Channels.access$000(Channels.java:61)
```
Same error as the one for firecracker.

fidencio commented:
- **jenkins-ci-ubuntu-18-04-inited**:
There are a bunch of failures there, including the "No space left on device", seen on both Firecracker and CLH-docker CIs.

amshinde commented:
@c3d Can you backport this PR to stable 1.12 as well?

amshinde commented:
I restarted 
jenkins-ci-ubuntu-1804-cloud-hypervisor-k8s-e2e-containerd-minimal job started by Fabiano, instead of the latest run. That has passed but the status not yet reported here. 

Ci run that passed:
http://jenkins.katacontainers.io/job/kata-containers-runtime-ubuntu-PR-cloud-hypeprvisor-k8s-e2e-containerd-minimal/174/

Finally vfio failure is unrelated and will be fixed with : https://github.com/kata-containers/tests/pull/3044

I think this PR is good to go, would like to see this backported to 1.12.

likebreath commented:
@amshinde @fidencio All required CIs are passing, and the PR is ready to go. Besides the known issue on `podman-clh-fedora-31`, do we want to land the PR with the failures on `opensuse-15` and `Power8-ubuntu-18-04-initrd`?
